### PR TITLE
SSCS-10135 Autolist Flag

### PR DESCRIPTION
**Pull request by TeaWithLucas**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10135

### Change description ###

- [x] Add Autolist conditions
- [x] Adjust panel specialisms
- [x] Add new tests for autolist
- [x] Adjust/add tests of changed methods

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java [hunk 1]
```diff
@@ -71,36 +70,37 @@ public static HearingDetails buildHearingDetails(HearingWrapper wrapper, Referen
...
+    public static LocalDate getHearingWindowStart(@Valid SscsCaseData caseData, boolean autoListed) {
 
-        if (!autoListed || isBlank(caseData.getDwpResponseDate())) {
-            LocalDate dateRangeStart = LocalDate.now().plusDays(DAYS_TO_ADD_HEARING_WINDOW_TODAY);
-            return HearingWindow.builder()
-                    .dateRangeStart(dateRangeStart)
-                    .dateRangeEnd(null)
-                    .build();
+        LocalDate startDate = LocalDate.now().plusDays(DAYS_TO_ADD_HEARING_WINDOW_TODAY);
+        if (isBlank(caseData.getDwpResponseDate())) {
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:88]: See my refactor of this for 10260. Basically, check if isNotBlank, then do DwpResponseDate stuff, then default at the end to LocalDate.now().plusDays(DAYS_TO_ADD_HEARING_WINDOW_TODAY)

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 2]
```diff
@@ -233,6 +211,24 @@ void testBuildHearingWindowNullReturn() {
...
+        SscsCaseData caseData = SscsCaseData.builder()
+                .dwpResponseDate("2021-12-01")
+                .urgentCase("Yes")
+                .build();
+        HearingWindow result = HearingsDetailsMapping.buildHearingWindow(caseData, false);
+
+        assertThat(result).isNotNull();
+
+        LocalDate expected = LocalDate.now().plusDays(DAYS_TO_ADD_HEARING_WINDOW_TODAY);
+        assertThat(result.getDateRangeStart()).isEqualTo("2021-12-15");
```
Comment (osheaf01) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:226]: Is expected being used?

**Review by osheaf01 — DISMISSED** [review]
LGTM now

**Review by nickwest1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java [hunk 3]
```diff
@@ -296,7 +300,11 @@ public static String getListingComments(Appeal appeal, List<CcdValue<OtherParty>
                     .collect(Collectors.toList()));
         }
 
-        return listingComments.isEmpty() ? null : String.join(String.format("%n%n"), listingComments);
+        if (listingComments.isEmpty()) {
+            return null;
+        }
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:305]: Why null, not a blank string?
If a blank string, then this if is not relevant too

**Review by nickwest1 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java (hunk 3 above)
Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:305]: This is an optional field on the request, if there are no comments, we do not want it to appear on the request so need to send null so the field isn't present on the request.

**Review by nickwest1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMapping.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static uk.gov.hmcts.reform.sscs.ccd.domain.YesNo.isYes;
+import static uk.gov.hmcts.reform.sscs.helper.mapping.HearingsCaseMapping.isInterpreterRequired;
+import static uk.gov.hmcts.reform.sscs.helper.mapping.HearingsCaseMapping.shouldBeAdditionalSecurityFlag;
+import static uk.gov.hmcts.reform.sscs.helper.mapping.HearingsMapping.getSessionCaseCode;
+
+public final class HearingsAutoListMapping {
```
Comment (wmarshid) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMapping.java:20]: minor: why are all these methods declared public, should they not be class protected at least? except the first one of course.

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                                        .build())
+                                .build())
+                        .hearingOptions(HearingOptions.builder()
+                                .build())
+                        .build())
+                .build();
+    }
+
+
+    @DisplayName("When there are not conditions that affect autolisting, shouldBeAutoListed returns true")
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java:44]: minor: typo "not" -> "no"

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                        false,false,SessionCategory.CATEGORY_01,null));
+
+        given(referenceData.getSessionCategoryMaps()).willReturn(sessionCategoryMaps);
+
+        boolean result = HearingsAutoListMapping.shouldBeAutoListed(caseData, referenceData);
+
+        assertThat(result).isTrue();
+    }
+
+    @DisplayName("When this is a condition that affects autolisting, shouldBeAutoListed returns false")
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java:58]: minor: typo "this" -> "there"

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    @Test
+    void testShouldBeAutoListedFalse() {
+        caseData.setLinkedCase(List.of(CaseLink.builder()
+                .value(CaseLinkDetails.builder()
+                        .caseReference("123456")
+                        .build())
+                .build()));
+        boolean result = HearingsAutoListMapping.shouldBeAutoListed(caseData, referenceData);
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java:66]: very minor: would put a new line above this

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+        assertThat(result).isFalse();
+    }
+
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java:70]: minor: extra new lines here

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    @DisplayName("When Representative is null, isRepresentativeOrg should return False")
+    @Test
+    void testIsRepresentativeOrgRepNull() {
+        boolean result = HearingsAutoListMapping.isRepresentativeOrg(null);
+
+        assertThat(result).isFalse();
+    }
+
+    @DisplayName("When hearingType is Paper, isPaperCaseAndNoPO return True")
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMappingTest.java:195]: minor/optional: if you want to change the method name remember to update these descriptions as well

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/ServiceHearingValuesMappingTest.java [hunk 10]
```diff
@@ -178,11 +183,16 @@ void shouldMapServiceHearingValuesSuccessfully() {
             "hearingLoop",
             "disabledAccess"
         ), serviceHearingValues.getFacilitiesRequired());
-        assertEquals(NOTE_FROM_OTHER_APPELLANT + NOTE_FROM_OTHER_APPELLANT + "\n" + "\n" + NOTE_FROM_OTHER_PARTY + NOTE_FROM_OTHER_PARTY, serviceHearingValues.getListingComments());
+        assertThat(serviceHearingValues.getListingComments())
+                .isEqualToNormalizingNewlines("Appellant - Mr Fred Flintstone:\n" + NOTE_FROM_APPELLANT
+                        + "\n\n" + "party_role - Mr Barny Boulderstone:\n" + NOTE_FROM_OTHER_PARTY);
```
Comment (wmarshid) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/ServiceHearingValuesMappingTest.java:188]: TIL: Barny's last name is Boulderstone!

**Review by wmarshid — DISMISSED** [review]
Good job mate!

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMapping.java [hunk 11]
```diff
@@ -29,7 +29,7 @@ public static boolean shouldBeAutoListed(@Valid SscsCaseData caseData, Reference
                 || shouldBeAdditionalSecurityFlag(caseData)
                 || isInterpreterRequired(caseData)
                 || HearingsDetailsMapping.isCaseLinked(caseData)
-                || isPaperCaseAndNoPo(caseData)
+                || isPaperCaseAndNoPoNotAttending(caseData)
```
Comment (wmarshid) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsAutoListMapping.java:32]: minor: can we change this from "NoPoNotAttending" -> "PoNotAttending" ?

**Review by wmarshid — APPROVED** [review]

**Review by wmarshid — APPROVED** [review]

**Review by nickwest1 — APPROVED** [review]

**Review by wmarshid — APPROVED** [review]

**Review by wmarshid — APPROVED** [review]

**Review by osheaf01 — COMMENTED** [review]
LGTM

**Pull request merged**